### PR TITLE
test: add randomized tests for errByzantine

### PR DESCRIPTION
**Pull request by walldiss**
ErrByzantine encompasses numerous edge cases. To address this, I implemented randomized fuzzing tests, which identified instances where shares encoding was not properly verified. 

While I am neutral about maintaining the pretty printing code (as it looks kinda off), I retained it because it proved extremely useful for understanding the reasons behind test failures during extensive debugging. Should these tests fail in the future for any reason, having a visual representation of the issues will expedite debugging for future developers.

Allowed to discover https://github.com/celestiaorg/rsmt2d/pull/313

### CODE REVIEW

File: extendeddatacrossword_test.go [hunk 1]
```diff
@@ -2,6 +2,7 @@ package rsmt2d
 
 import (
 	"bytes"
+	crand "crypto/rand"
 	"errors"
 	"fmt"
 	"math/rand"
```
Comment (rootulp) [inline extendeddatacrossword_test.go:8]: [no change needed] in a follow up PR we may consider replacing usage of `math/rand` with `crypto/rand` and then deleting the import alias for `crypto/rand`

File: extendeddatacrossword_test.go [hunk 2]
```diff
@@ -423,6 +424,186 @@ func TestCorruptedEdsReturnsErrByzantineData_UnorderedShares(t *testing.T) {
...
+	t.Skip()
+	for i := 0; i < 10000; i++ {
+		TestErrRandByzantine(t)
+	}
+}
+
+func TestErrRandByzantine(t *testing.T) {
+	codec := NewLeoRSCodec()
+	original, corrupted, idx := randCorruptedEDS(t, codec, 8)
+	require.False(t, original.Equals(corrupted), "corrupted eds is equal to original eds")
```
Comment (rootulp) [inline extendeddatacrossword_test.go:438]: [question] why `.False` instead of:

```go
	require.NotEqual(t, original, corrupted)
```

File: extendeddatacrossword_test.go [hunk 3]
```diff
@@ -423,6 +424,186 @@ func TestCorruptedEdsReturnsErrByzantineData_UnorderedShares(t *testing.T) {
...
+		shareSize,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failure to create extended data square: %w", err)
+	}
+
+	// set corrupted share first
+	errX, errY := corruptedIdx/int(corrupted.Width()), corruptedIdx%int(corrupted.Width())
+	share := corrupted.GetCell(uint(errX), uint(errY))
+	err = square.SetCell(uint(errX), uint(errY), share)
```
Comment (rootulp) [inline extendeddatacrossword_test.go:496]: [nit] perhaps avoid `errX` and `errY` because it's easy to visually conflate them with `err`. Candidates: `corruptedX` and `corruptedY`?

File: extendeddatacrossword_test.go [hunk 4]
```diff
@@ -423,6 +424,186 @@ func TestCorruptedEdsReturnsErrByzantineData_UnorderedShares(t *testing.T) {
...
+	if err != nil {
+		return nil, fmt.Errorf("failure to get column roots: %w", err)
+	}
+
+	// loop until repaired or byzantine error
+	for {
+		repaired, err := fillRandomCellAndRepair(corrupted, square, rowRoots, colRoots, samples)
+		if repaired {
+			prettyPrintSamples(samples, corruptedIdx)
+			return square, errors.New("no byzantine error")
```
Comment (rootulp) [inline extendeddatacrossword_test.go:515]: [question] why return an error her? Why not

```suggestion
			return square, nil
```

**Review by rootulp — APPROVED** [review]
LGTM

File: extendeddatacrossword_test.go (hunk 4 above)
Comment (walldiss) [reply extendeddatacrossword_test.go:515]: Test corrupts eds and tries to repair it. It shoud result in Byzantine every time on corrupted eds. If there was no encoding errors during eds construction, it means we missed some cases.

File: extendeddatacrossword_test.go (hunk 3 above)
Comment (walldiss) [reply extendeddatacrossword_test.go]: sgtm https://github.com/celestiaorg/rsmt2d/pull/312/commits/61f82d1333a6bd5d3e483b56f108ba00d70f1c16

File: extendeddatacrossword_test.go (hunk 1 above)
Comment (walldiss) [reply extendeddatacrossword_test.go:8]: Unfortunately, `crypto/rand` lacks handy IntN funciton, only 
```
func Int(rand io.Reader, max *big.Int) (n *big.Int, err error) 
```
While `math/rand` lacks ability to generate random []byte slice. I wish one day there will be single package for both

File: extendeddatacrossword_test.go (hunk 2 above)
Comment (walldiss) [reply extendeddatacrossword_test.go:438]: NotEqual uses `==` slice compare under the hood, while `Equals` method does more precise comparison.

File: extendeddatacrossword_test.go (hunk 1 above)
Comment (Wondertan) [reply extendeddatacrossword_test.go:8]: There is rand v2

**Review by Wondertan — APPROVED** [review]

**Review by evan-forbes — APPROVED** [review]
